Docs - complete v1.2.0 documentation w/ secret store & keyvault overriden (#134)
arcus-azure/arcus.security: docs/preview/features/secrets/secret-store-integration.md

> ---

Reviewer: More input can be pulled from the blog post when it's ready but this is good for now.

> ---

Author: Added a proposal.

> ---

Author: Added a proposal.

title: "Secret store integration"
layout: default
---

> # Secret Store Integration

Reviewer: ```suggestion
# Using Secret Store
```

As alternative to the usage of placing secrets into an `IConfiguration` instance in your application, the `Arcus.Security.Core` package provides a alternative concept called 'secret stores'.

> The setup is simulair as to building an `IConfiguration` instance, by adding the different 'secret store sources' where the secrets should be present.

Reviewer: How about this? Feels more smooth or not?
```suggestion
We provide an approach similar to how `IConfiguration` is built, but with a focus on secrets. You can pick and choose the secret providers you want to use and we'll get the job done!
```

> All of those sources combined, creates then a single `ISecretProvider` that the application can use and has access to all the different registered secret sources.

Reviewer: Just a suggestion, but not a must:
```suggestion
Once register, you can fetch all secrets by using `ISecretProvider` which will get secrets from all the different registered secret providers.
```

## Installation
For this feature, the following package needs to be installed:

```shell
PM > Install-Package Arcus.Security.Core
```

## Usage
The secret stores are configured during the initial application build-up in the `Program.cs`:

```csharp
public class Program
{
    public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureAppConfiguration((context, config) => 
                {
                    config.AddJsonFile("appsettings.json")
                          .AddJsonFile("appsettings.Development.json");
                })
                .ConfigureSecretStore((context, config, builder) =>
                {
                    builder.AddEnvironmentVariables();
                })
                .ConfigureWebHostDefaults(webBuilder => webBuilder.UseStartup<Startup>());
    }
}
```

> After the registration, the `ISecretProvider` can be used as any other registered service:

Reviewer: ```suggestion
Once the secret providers are defined, the `ISecretProvider` can be used as any other registered service:
```

```csharp
[ApiController]
public class HealthController : ControllerBase
{
    public HealthController(ISecretProvider secretProvider)
    {
    }
}
```

## Built-in secret sources

> Several built in secret sources available in the package.

Reviewer: I would just list them all here and point to dedicated pages:
```suggestion
Several built-in secret sources available in the package:

- [Azure Key Vault]()
- [Configuration]()
- [Environment Variables]()
```

Which are alphabetically ordered

Author: Yes, that's done.

If you require an additional secret sources that aren't available here, please [this document](./create-new-secret-source) that describes how you can create your own secret soure.

> **Environment variables**

Reviewer: I would move these to dedicated pages as well

Author: Yes, that's done.

Using the environment variable secret source, the secrets will be searched in the the variables on the evnvironment.

```csharp
.ConfigureSecretStore((context, config, builder) =>
{
    builder.AdEnvironmentVariables();
})
```

> **IConfiguration**

Reviewer: I would move these to dedicated pages as well

Author: Hmm, isn't that a lot of back-and-forth?

Author: Is done.
